rust_src/src/multibyte.rs

```diff
@@ -110,6 +110,35 @@ impl LispStringRef {
     pub fn byte_at(&self, index: ptrdiff_t) -> u8 {
         unsafe { *self.const_data_ptr().offset(index) }
     }
+
+    /// (Somewhat) Replaces STRING_SET_UNIBYTE in C. That macro has the behavior that
+    /// if your string has size 0, it will replace your string variable with
+    /// 'empty_unibyte_string'. This function does not have that side-effect.
+    #[inline]
+    pub fn set_unibyte(mut self) {
+        if self.size != 0 {
+            self.size_byte = -1;
```

Comment:
This is changing the behaviour, isn't it?

If we're no longer replacing an empty string, then elisp code could change behaviour because it could observe whether the string pointer has changed.

Reply:
I don't believe it should change behavior based on that macro definition. Basically, in the case that size == 0, it would assign `empty_string_unibyte` to the C layer Lisp_Object. Since SET_STRING_UNIBYTE(...) is the last call before we return Qnil, that side effect shouldn't matter. BUT it could change the behavior if people use this function in place of other places we use SET_STRING_UNIBYTE in the future, if they think this function is a 1:1 replacement. 

``` c
/* Mark STR as a unibyte string.  */
#define STRING_SET_UNIBYTE(STR)				\
  do {							\
    if (XSTRING (STR)->size == 0)			\
      (STR) = empty_unibyte_string;			\
    else						\
      XSTRING (STR)->size_byte = -1;			\
  } while (false)
```

Reply:
Why are you not emptying the string?

Reply:
In the case of `clear_string`, it doesn't matter, since we don't use STR after calling STRING_SET_UNIBYTE. The case that size == 0 doesn't mutate the underlying data of the untagged pointer, so it doesn't change the behavior.

Eh honestly I'm just not in love with what this macro is doing semantically, and this was an attempt to move away from it. In one case it sets a variable on your untagged pointer, in the other case it changes what your pointer is pointer to. 

I'm going to make an update to provide a function that will provide an inline function that does something similar to this macro. I think it will help avoid confusion and avoid errors in the future.